### PR TITLE
Fix the Engine version check in GUI

### LINKED ISSUE #6560: Project created with a new nightly package is requesting an outdated engine
### Discord username

_No response_

### What type of issue is this?

Permanent – Occurring repeatably

### Is this issue blocking you from using Enso?

- [ ] Yes, I can't use Enso because of this issue.

### Is this a regression?

- [X] Yes, previous version of Enso did not have this issue.

### What issue are you facing?

IDE is reporting unsupported Engine version for a project just created with newest nightly.

### Expected behaviour

No warning - I have newer version then the one required

### How we can reproduce it?

_No response_

### Screenshots or screencasts

image

### Enso Version

nightly 04.05

### Browser or standalone distribution

Standalone distribution

### Browser Version or standalone distribution

standalone

### Operating System

MacOS

### Operating System Version

_No response_

### Hardware you are using

_No response_

### PR DESCRIPTION
**Pull request by mwu-tow**
### Pull Request Description
This PR fixes #6560. 

The fix has a few elements:
1) Bumps the Engine requirement to the latest release, namely `2023.1.1`.
2) Changed the logic of checking whether a given version matches the requirement. Previously, we relied on `VersionReq` from `semver` crate which did not behave intuitively when the required version had a prerelease suffix. Now we rely directly on Semantic Versioning rules of precedence.
3) Code cleanups, including deduplicating 3 copies of the version-checking code, and moving some tests to more sensible places.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

File: app/gui/config/src/lib.rs [hunk 1]
```diff
@@ -20,17 +20,62 @@ use enso_json_to_struct::json_to_struct;
 
 
 // ==============
-// === Config ===
+// === Errors ===
 // ==============
 
+/// A wrapper for an error with information that the Engine version does not meet the requirements.
```
Comment (kazcw) [inline app/gui/config/src/lib.rs]: Is this a wrapper for an error or simply an error type?

File: app/gui/config/src/lib.rs [hunk 2]
```diff
@@ -20,17 +20,62 @@ use enso_json_to_struct::json_to_struct;
...
+#[derive(Clone, Debug, thiserror::Error)]
+#[error("Unsupported Engine version: required {required} (or newer), found {found}.")]
+pub struct UnsupportedEngineVersion {
+    /// The version of the Engine that is required.
+    pub required: semver::Version,
+    /// The version of the Engine that was found.
+    pub found:    semver::Version,
+}
+
```
Comment (kazcw) [inline app/gui/config/src/lib.rs:35]: More whitespace.

File: app/gui/config/src/lib.rs [hunk 3]
```diff
@@ -20,17 +20,62 @@ use enso_json_to_struct::json_to_struct;
...
+///
+/// Effectively, this checks if the given version is greater or equal to the minimum supported.
+/// "Greater or equal" is defined by the [Semantic Versioning specification](https://semver.org/)
+/// term of precedence.
+pub fn check_engine_version(
+    engine_version: &semver::Version,
+) -> Result<(), UnsupportedEngineVersion> {
+    // We don't want to rely on the `semver::VersionReq` semantics here. Unfortunately the
+    // [Semantic Versioning specification](https://semver.org/) does not define the semantics of
+    // of the version requirement operators, so different implementations may behave differently.
```
Comment (kazcw) [inline app/gui/config/src/lib.rs]: Doubled `of`

File: app/gui/config/src/lib.rs [hunk 4]
```diff
@@ -20,17 +20,62 @@ use enso_json_to_struct::json_to_struct;
...
+    //
+    // The `semver::VersionReq` implementation follows the Cargo's implementation, namely:
```
Comment (kazcw) [inline app/gui/config/src/lib.rs:65]: ~~the~~ Cargo

File: app/gui/config/src/lib.rs [hunk 5]
```diff
@@ -20,17 +20,62 @@ use enso_json_to_struct::json_to_struct;
...
+    // ```
+    // In particular, in order for any VersionReq to match a pre-release version, the VersionReq
+    // must contain at least one Comparator that has an explicit major, minor, and patch version
+    // identical to the pre-release being matched, and that has a nonempty pre-release component.
+    // ```
```
Comment (kazcw) [inline app/gui/config/src/lib.rs:70]: Is there a source you can readily link for this?

File: app/gui/config/src/lib.rs [hunk 6]
```diff
@@ -64,3 +109,24 @@ pub fn read_args() -> Args {
 lazy_static! {
     pub static ref ARGS: Args = read_args();
 }
+
```
Comment (kazcw) [inline app/gui/config/src/lib.rs:126]: Missing vertical space.

File: app/gui/config/src/lib.rs [hunk 7]
```diff
@@ -64,3 +109,24 @@ pub fn read_args() -> Args {
...
+        // We just expect that it won't panic.
+        let _ = engine_version_required();
+    }
+
+    #[test]
+    fn new_project_engine_version_fills_requirements() {
+        // Sanity check: required version must be supported.
+        assert!(check_engine_version(&engine_version_required()).is_ok());
+    }
+}
```
Comment (kazcw) [inline app/gui/config/src/lib.rs:159]: What about a test that would fail with `semver::VersionReq`?

**Review by kazcw — APPROVED** [review]
The previous error message gave instructions to modify a project to use a different engine version, while the new message does not; I assume this is an intentional change.

**Comment by wdanilo** [comment]
I don't think this is intentional to not report users with the error that Kaz mentioned above. @mwu-tow ?

**Comment by wdanilo** [comment]
Feel free to merge it after fixing the issues. If there is Kaz review, there is no need for mine :)

**Comment by mwu-tow** [comment]
@wdanilo 
I'm not sure. I felt that this warning was not making much sense, especially if the primary use case we target for our clients is a cloud-based project setup. In such cases, the warning gives wrong information, the file is not on the disk, and path on the remote would be totally different (if it was accessible).

**Pull request merged**